### ISSUE TITLE
Crypt support for public/private keys (public key crypto)

### ISSUE THREAD

**Issue by lanrat** [ncw/rclone#737]
This is less of a bug and more of a feature request/discussion. If this is the wrong place for this then feel free to close and point me to the correct place.

In addition to using a password+salt for the Crypt remote, it would be useful if rclone could encrypt/decrypt with public/private key pairs. This would allow rclone to perform an encrypted backup to a remote location without the need for the system rclone is running on to decrypt the backups.

For example, if your system is compromised and an attacker gets access to the rclone.conf file they could decrypt your remote backups, but if they were encrypted using a public key and the private key was not on the system the data would be useless. Only the user with the private key would be able to decrypt the backups.

This would still require rclone to be able to access metadata about the files, so this should be used for file contents only.

**Comment by ncw** [comment]
I think this is a good idea. To encrypt a file, you'd make a session key, encrypt that with the public key then store that blob in the file header.  To decrypt you'd read the blob out of the file header, decrypt with private key then use the session key to decrypt the rest of the message.

**Comment by lanrat** [comment]
@ncw do you plan to use a new random session key for each file? (You should).

Also, you need to be sure that session keys are not stored anywhere in the system, including in the logs as it would be able to decrypt any of the files without the need for the public or private key.

**Comment by geek-merlin** [comment]
Related / duplicated by #1295

**Comment by geek-merlin** [comment]
Maybe useful for this: https://github.com/str4d/rage

**Comment by VictorNine** [comment]
Related / duplicated by #4292

**Comment by VictorNine** [comment]
I've started hacking this together. I've copied the crypt back end and made some modifications. I create a 32 byte random session key which is stored in the file header. This is to use symmetric encryption for every file. Right now this is stored in plain text. Next step is to encrypt this with a public key. So the question is how this should be encrypted I see two options.

1. Integrate with GPG on the system and let the session key be encrypted there. Guessing this will cause some issues between Windows and Linux.

2. Handle it inside rclone where the private/public key information is stored in the config file.

Thoughts? I'm leaning towards option 2

**Comment by geek-merlin** [comment]
@VictorNine Awesome!💪

In the end i guess we want to pluginify this and choose between these and meybe more options. So while i like 1 for reasons outlined in #4292, i'd encourage to first take the shortest path to get that POC out of the garage, which is probably 2.

**Comment by VictorNine** [comment]
I now have a working POC for option 2. (Hacked together but working). I'm using RSA which adds 128bytes to every file. This is to store the encrypted session key.

I think this should be it's own back end not baked in to crypt. Opinions?

**Comment by ncw** [comment]
> I now have a working POC for option 2. (Hacked together but working). I'm using RSA which adds 128bytes to every file. This is to store the encrypted session key.
> 
> I think this should be it's own back end not baked in to crypt. Opinions?

If you check out https://github.com/rclone/rclone/issues/3667 you'll see some ideas for adding metadata to crypt files.

Perhaps this could fit in at the same time. Presumably your 128bytes extra is variable depending on the length of the RSA key?

If we are changing the crypt format then I would like to use a new session key for each file - this would then enable people to share crypted files without having to share the master key.

**Comment by VictorNine** [comment]
Thanks, looks that is still an open discussion. For now I created a new back end where the header is extended by 128 bytes. As you assumed this will vary depending on key length. So for now it's lock to 1024 bit (should be fine for most uses).

In the future maybe this can be merged into crypt with more header data. Would also be nice to be able to have multiple "recipients" for the file.

**Comment by VictorNine** [comment]
You can see my suggestion here https://github.com/VictorNine/rclone/tree/pubcrypt

There is one issue remaining before I can send a pull request. You have the option to create a new key pair in config. That sets the private key. The issue is how I should show the public key. Any suggestions?

**Comment by ncw** [comment]
> Thanks, looks that is still an open discussion. For now I created a new back end where the header is extended by 128 bytes. As you assumed this will vary depending on key length. So for now it's lock to 1024 bit (should be fine for most uses).
> 
> In the future maybe this can be merged into crypt with more header data. Would also be nice to be able to have multiple "recipients" for the file.

I'd like to put into place the crypt header extension before merging something like this into the crypt backend.

> You can see my suggestion here https://github.com/VictorNine/rclone/tree/pubcrypt

Nice!

> There is one issue remaining before I can send a pull request. You have the option to create a new key pair in config. That sets the private key. The issue is how I should show the public key. Any suggestions?

We'll need to keep both private key and the public key in the config file won't we?

Is there a standard way of showing these keys?

---

I note also that rclone crypt is quite similar to age encryption: https://github.com/FiloSottile/age

Maybe we would be better re-using that to make a new crypt? It has variable length headers which would be a bit of a problem but maybe they could be padded?

**Comment by VictorNine** [comment]
> I'd like to put into place the crypt header extension before merging something like this into the crypt backend.
Yes that should probably be agreed first. Would you like to merge in a new back end like I have implemented it? At least it provides the feature. If it's kept as two commits it's fairly easy to see the difference to crypt.

> We'll need to keep both private key and the public key in the config file won't we?
> Is there a standard way of showing these keys?
The public key is derived from the private key. So you only need the private key in the config. If you leave that blank you need to provide a public key. The issue is that I cant figure out how to write to both "fields" when the key is created. The other option is to have an export flag.

> I note also that rclone crypt is quite similar to age encryption: https://github.com/FiloSottile/age
> 
> Maybe we would be better re-using that to make a new crypt? It has variable length headers which would be a bit of a problem but maybe they could be padded?
The issue with this as I see it is that a 1024 bit key adds 128 bits to the header, a 2048 bit key adds 256 bits to the header and so on. If you want multiple recipients you need to multiply that by max recipients. Which means your header will be "huge". To make that work you need to have a variable length header. Not familiar with age but I'll look in to it.
